### ISSUE TITLE
Recursive use of block_on panics

### ISSUE THREAD

**Issue by That3Percent** [rust-lang/futures-rs#2090]
When `futures::executor::block_on` is called recursively it panics with:

```
thread 'main' panicked at 'cannot execute `LocalPool` executor from within another executor: EnterError', src/libcore/result.rs:1188:5
```

Here is a complete program to reproduce this result:

```rust
fn main() {
    use futures::executor::block_on;
    block_on(async { block_on( async {} ) })
}
```

The documentation for [block_on](https://docs.rs/futures/0.3.4/futures/executor/fn.block_on.html) does not mention any limitation that would cause a panic, and it is unclear (to me at least) that such a limitation would be necessary.

The actual code that causes this panic is here in [local_pool.rs](https://github.com/rust-lang/futures-rs/blob/56980e76b9eb9827d56cdbbeaf28a2f018a89a16/futures-executor/src/local_pool.rs#L78)

**Comment by yoshuawuyts** [comment]
I believe our implementation of block on in async-std handles this case: [impl here](https://github.com/async-rs/async-std/blob/master/src/task/block_on.rs). Maybe it could make sense for futures-rs to adopt a similar fix?

**Comment by seanmonstar** [comment]
It purposefully panics because a recursive call means blocking any other futures that were running on the same thread.

**Comment by cramertj** [comment]
Yeah, @seanmonstar is right-- this is "working as intended". It's quite bugprone (or arguably always wrong) to block inside the implementation of a future. I'll leave this issue open for discussion, but I don't personally think there's a bug to be fixed here.

**Comment by That3Percent** [comment]
Yes, I'm coming around and think that the panic here is reasonable and desirable.

I was originally thinking that at the very least the docs should be amended to document the potential panic - but the more I think about it "Future.poll must block" is documented and must be well-known and adhered to across the ecosystem. Violating that constraint could cause anything to happen (undefined behavior excepted) in library code - even panics.

**Comment by kevincox** [comment]
I have a use case where I know that the parent executor is idle and need to interact with some code that isn't async. It would be nice to be able to do this for the weird cases where it is needed. I know that it isn't the "perfect" solution in most cases but when you need it, well, you need it.

**Comment by b00f** [comment]
This example also panics with the same message:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3147b7fe8a6938fd0e3da2a7dc8b7209

**Comment by tgiddings** [comment]
If block_on purposely panics, it should give a relevent panic message. The current one is phrased in terms of the internals of block_on, which is not informative and implies that the panic is not correct.

**Comment by taiki-e** [comment]
We can allow this once the lint like the one that [this shiny future story](https://rust-lang.github.io/wg-async-foundations/vision/shiny_future/barbara_enjoys_an_async_sandwich.html) mentions is implemented.

**Comment by FlorianUekermann** [comment]
> Yeah, @seanmonstar is right-- this is "working as intended". It's quite bugprone (or arguably always wrong) to block inside the implementation of a future. I'll leave this issue open for discussion, but I don't personally think there's a bug to be fixed here.

What about cases where the future doesn't actually block or is guaranteed to finish on its own?

I have called `block_on` in cases where I use an async api with something like `Cursor<Vec<u8>>` as `AsyncWrite`, which shouldn't block. Now I'm learning that this may panic if a user happens to call my code within async code. That's pretty bad.

This makes `block_on` unsuitable for any library code and the documentation contains no mention of it.

I guess this is the correct solution in some cases, but that may be a bit too difficult to come up with for someone who has nothing to do with async and just wants to use an async API in their mostly non-async world.
```
        let mut ctx = Context::from_waker(noop_waker_ref());
        match my_future.poll_unpin(&mut ctx) {
            Poll::Ready(r) => r,
            Poll::Pending => unreachable!(),
        }
```
Not sure what the correct solution is cases where transient blocking may happen.

**Comment by chessai** [comment]
This seems to affect quickcheck_async when using the filesystem or Cursor mocks: https://docs.rs/quickcheck_async/latest/src/quickcheck_async/lib.rs.html#73-119
